### PR TITLE
sample in O(limit)

### PR DESCRIPTION
**Pull request by generall**
The suggested algorithm:
- guarantees to finish in O(limit)  - assuming the selected points are unique
- subjectively easier to understand

### CODE REVIEW

File: lib/collection/src/shards/local_shard/scroll.rs [hunk 1]
```diff
@@ -280,51 +281,57 @@ impl LocalShard {
...
-
-            if !rng.gen_bool(*weight) {
-                continue;
+        let mut random_points = HashSet::with_capacity(limit);
+
+        // This loop iterates <= LIMIT times, and either quits if we have enough points,
+        // or if some of the segments are exhausted.
+        //
+        // If the segments are exhausted, we will fill up the rest of the points from other segments.
+        // In total, the complexity is guaranteed to be O(limit).
```
Comment (timvisee) [inline lib/collection/src/shards/local_shard/scroll.rs:303]: I feel like the two parts contradict each other:
- the first says it stops if a segment is exhausted
- the second says we'll fill up with other segments

Seeing the implementation I believe this is intentional. Maybe this helps clarify a bit though:

```rust
        // Randomly sample points in two stages
        //
        // 1. This loop iterates <= LIMIT times, and either breaks early if we
        // have enough points, or if some of the segments are exhausted.
        //
        // 2. If the segments are exhausted, we will fill up the rest of the
        // points from other segments. In total, the complexity is guaranteed to
        // be O(limit).
```

File: lib/collection/src/shards/local_shard/scroll.rs [hunk 2]
```diff
@@ -280,51 +281,57 @@ impl LocalShard {
...
-            };
-
-            // Add point to result
-            random_points.push(point);
-            if random_points.len() == limit {
-                break;
             }
         }
 
+        let selected_points: Vec<_> = random_points.into_iter().collect();
```
Comment (timvisee) [inline lib/collection/src/shards/local_shard/scroll.rs:330]: Is it possible for the output to be somewhat deterministic here due to the hash set? I honestly don't know.

If it is, I think that's undesirable.

**Review by timvisee — APPROVED** [review]

File: lib/collection/src/shards/local_shard/scroll.rs (hunk 2 above)
Comment (timvisee) [reply lib/collection/src/shards/local_shard/scroll.rs:330]: Assuming each set has an unique [`RandomState`](https://doc.rust-lang.org/std/collections/hash_map/struct.RandomState.html), it does not have this problem.

**Review by coszio — APPROVED** [review]
I agree this easier to reason about. Thanks.

**Pull request merged**